recipes-azure-iot/iot-hub-device-update/iot-hub-device-update/swupdate_handler_v2_grub.sh

```diff
@@ -0,0 +1,757 @@
+}
+
+#
+# Performs steps to finalize the update.
+#
+ApplyUpdate() {
+    log_info "Applying update..."
+
+    ret_val=0
+    resultCode=0
+    extendedResultCode=0
+    resultDetails=""
+
+    echo "Applying." >> "${log_file}"
+
+    if [ -f "/run/omnect-bootloader-update" ]; then
+        bootloader_env.sh set omnect_os_bootpart $update_part
+        echo "use omnect_os_bootpart environment" >> "${log_file}"
+    else
+        bootloader_env.sh set omnect_validate_update_part $update_part
+        echo "use omnect_validate_update_part environment" >> "${log_file}"
+    fi
+    ret_val=$?
+
+    if [[ $ret_val -eq 0 ]]; then
+
+        log_info "restart_to_apply=$restart_to_apply"
+        log_info "restart_agent_to_apply=$restart_agent_to_apply"
+
+        if [[ ${restart_to_apply} == "yes" ]]; then
+            log_info "Returning ADUC_Result_Apply_RequiredImmediateReboot(705)"
+            resultCode=705
+        elif [[ ${restart_agent_to_apply} == "yes" ]]; then
+            log_info "Returning ADUC_Result_Apply_RequiredImmediateAgentRestart(707)"
+            resultCode=707
+        else
+            log_info "Returning ADUC_Result_Apply_Success(700)"
+            resultCode=700
+        fi
+        extendedResultCode=0
+        resultDetails=""
+    else
+        resultCode=0
+        make_swupdate_handler_erc "$ret_val" extendedResultCode
+        resultDetails="Cannot set update partition."
+        log_error "$resultDetails"
+    fi
+
+    # Prepare ADUC_Result json.
+    aduc_result=""
+    make_aduc_result_json "$resultCode" "$extendedResultCode" "$resultDetails" aduc_result
+
+    # Show output.
+    output "Result:" "$aduc_result"
+
+    # Write ADUC_Result to result file.
+    result "$aduc_result"
+
+    $ret $ret_val
+}
+
+#
+# Cancel current update.
+#
+CancelUpdate() {
+    log_info "CancelUpdate called"
+
+    echo "Cancelling update." >> "${log_file}"
+    resultCode=0
+    extendedResultCode=0
+    resultDetails=""
+    ret_val=
+
+    echo "Revert update." >> "${log_file}"
+    bootloader_env.sh unset omnect_validate_update_part
+    ret_val=$?
+
+    if [[ $ret_val -eq 0 ]]; then
+        resultCode=800
+    else
+        resultCode=801
+        make_swupdate_handler_erc "$ret_val" extendedResultCode
+        resultDetails="Failed to unset omnect_validate_update_part"
+        ret_val=0
+    fi
+
+    make_aduc_result_json "$resultCode" "$extendedResultCode" "$resultDetails" aduc_result
+
+    # Show output.
+    output "Result:" "$aduc_result"
+
+    # Write ADUC_Result to result file.
+    result "$aduc_result"
+
+    $ret $ret_val
+}
+
+#
+# Main
+#
+if [ -n "$check_is_installed" ]; then
+    CheckIsInstalledState "$installed_criteria"
+    exit $ret_val
+fi
+
+if [ -n "$do_download_action" ]; then
+    DownloadUpdateArtifacts
+    exit $ret_val
+fi
+
+if [ -n "$do_install_action" ]; then
+    InstallUpdate
+    exit $ret_val
+fi
+
+if [ -n "$do_apply_action" ]; then
+    ApplyUpdate
+    exit $ret_val
```

Comment:
i can theoretically use both parameter in one step ... `--action-install` and `--action-apply` but only the install would be applied and the script would return with the result code of the install step. 
Same finding for other usage of multiple other parameters as well.

Reply:
imho, the script will only be used in the automated environment like swupdate_handler_v2 or in the smoke-tests...